pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java

```diff
@@ -456,19 +457,23 @@ public synchronized void purgeInactiveProducers() {
                 .toMillis(pulsar.getConfiguration().getBrokerDeduplicationProducerInactivityTimeoutMinutes());
 
         Iterator<java.util.Map.Entry<String, Long>> mapIterator = inactiveProducers.entrySet().iterator();
+        boolean hasInactive = false;
         while (mapIterator.hasNext()) {
             java.util.Map.Entry<String, Long> entry = mapIterator.next();
             String producerName = entry.getKey();
             long lastActiveTimestamp = entry.getValue();
 
-            mapIterator.remove();
-
             if (lastActiveTimestamp < minimumActiveTimestamp) {
                 log.info("[{}] Purging dedup information for producer {}", topic.getName(), producerName);
+                mapIterator.remove();
                 highestSequencedPushed.remove(producerName);
                 highestSequencedPersisted.remove(producerName);
+                hasInactive = true;
             }
         }
+        if (hasInactive) {
+            takeSnapshot(getManagedCursor().getMarkDeletedPosition());
```

Comment:
If there are a large number of topics, snapshots may be created concurrently. We should avoid creating snapshots frequently. This is now done by the timer thread.

Reply:
`checkMessageDeduplicationInfo` is not a frequent operation, but it traverses all topics each time when it is executed. If many topics have Producers that need to be purged, a large number of snapshots will be generated.
Currently, `takeSnapshot` is automatically executed every 120 seconds. Therefore, we do not need to generate a snapshot after each topic is purged.

Reply:
Thanks your reply, the main reason I triggered a snapshot in the `purgeInactiveProducers` method was. If we rely on `deduplication-snapshot-monitor` timer to take snapshot, that may not be executed. Because we did the optimizations in the `takeSnapshot()` method.
```
        PositionImpl markDeletedPosition = (PositionImpl) managedCursor.getMarkDeletedPosition();
        if (markDeletedPosition != null && position.compareTo(markDeletedPosition) <= 0) {
            return;
        }
```

When `dup.cursor.markDeletedPosition == ledger.lastConfirmedEntry` and all producers stoped. In this case, we will not be able to clean up these producers. Unless new data is written and take a snapshot. The can be confusing for users.

> but it traverses all topics each time when it is executed. If many topics have Producers that need to be purged, a large number of snapshots will be generated.

In effect, automatically `takeSnapshot` also traverses all topics.

> we do not need to generate a snapshot after each topic is purged. Or, wait until all topics are cleaned up and create a snapshot.

This PR is only take a snapshot when there are inactive topics

In conclusion, the snapshot here is to solve the problem described above, `checkMessageDeduplicationInfo` is not a frequent operation.

Thanks again, If you have any suggestions, please continue the discussion.